services/spar/test-integration/Test/Spar/APISpec.hs

```diff
@@ -638,21 +638,97 @@ specCRUDIdentityProvider = do
           (owner, _, (^. idpId) -> idpid) <- registerTestIdP
           callIdpUpdate' (env ^. teSpar) (Just owner) idpid (IdPMetadataValue "<NotSAML>bloo</NotSAML>" undefined)
             `shouldRespondWith` ((== 400) . statusCode)
-    describe "issuer changed to one that already exists in another team" $ do
+    describe "issuer changed to one that already exists in *another* team" $ do
       it "rejects" $ do
         env <- ask
         (owner1, _, (^. idpId) -> idpid1) <- registerTestIdP
         (_, _, _, (IdPMetadataValue _ idpmeta2, _)) <- registerTestIdPWithMeta
         callIdpUpdate' (env ^. teSpar) (Just owner1) idpid1 (IdPMetadataValue (cs $ SAML.encode idpmeta2) undefined)
-          `shouldRespondWith` checkErr (== 400) "idp-used-in-other-team"
-    describe "issuer changed to one that is new" $ do
+          `shouldRespondWith` checkErr (== 400) "idp-issuer-in-use"
+    describe "issuer changed to one that already exists in *the same* team" $ do
       it "rejects" $ do
         env <- ask
-        (owner, _, (^. idpId) -> idpid, (IdPMetadataValue _ idpmeta, _)) <- registerTestIdPWithMeta
-        newissuer <- makeIssuer
-        let idpmeta' = idpmeta & edIssuer .~ newissuer
-        callIdpUpdate' (env ^. teSpar) (Just owner) idpid (IdPMetadataValue (cs $ SAML.encode idpmeta') undefined)
-          `shouldRespondWith` checkErr (== 400) "cannot-update-idp-issuer"
+        (owner1, _, (^. idpId) -> idpid1, (IdPMetadataValue _ idpmeta1, _)) <- registerTestIdPWithMeta
```

Comment:
This... took me at least a 1 minute to parse in my head haha.

Reply:
Personally; I'd like if all these helper functions would use properly named datatypes instead of these nested tuples. This is very hard to read without context.

Reply:
Yes.  I think I have a FUTUREWORK about that somewhere.